### ISSUE TITLE
Dart compiler crashes for unknown reason

### ISSUE THREAD

**Issue by ndevir** [dart-lang/sdk#45349]
Hello!

First of all, thanks a lot for all your efforts - I think dart is really great!

Now, on to the problem: The dart compiler crashed when I tried to compile the following code:

```
String parseColumnItem(List row, String columnName, Map columnPositions){
  int columnPosition = columnPositions[columnName];
  if(columnPosition == -1 or columnPosition == null){
    return null
  }
  return row.elementAt(columnPosition).toString();
}
```

I know the code is full of errors (and I have fixed it), so that's not why I'm submitting this issue. The point is that the error message told me to submit the issue, so here we go.

This was the console output:

```
Error compiling to JavaScript:
Warning: Interpreting this as package URI, 'package:dartpad_sample/main.dart'.
lib/main.dart:3:27:
Error: Expected ')' before this.
  if(columnPosition == -1 or columnPosition == null){
                          ^^
lib/main.dart:
Internal Error: The compiler crashed when compiling this element.

The compiler is broken.

When compiling the above element, the compiler crashed. It is not
possible to tell if this is caused by a problem in your program or
not. Regardless, the compiler should not crash.

The Dart team would greatly appreciate if you would take a moment to
report this problem at http://dartbug.com/new.

Please include the following information:

* the name and version of your operating system,

* the Dart SDK build number (2.12.1), and

* the entire message you see here (including the full stack trace
  below as well as the source location above).

The compiler crashed: Crash when compiling file:///tmp/dartpadCQODDS/lib/main.dart,
at character offset 71:
Null check operator used on a null value
#0      Parser._attemptPrecedenceLevelRecovery (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4901:77)
#1      Parser._parsePrecedenceExpressionLoop (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4887:11)
#2      Parser.parsePrecedenceExpression (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4754:12)
#3      Parser.parseExpression (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4680:13)
#4      Parser.parseReturnStatement (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4587:13)
#5      Parser.parseStatementX (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4454:14)
#6      Parser.parseStatement (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4431:20)
#7      Parser.parseBlock (package:_fe_analyzer_shared/src/parser/parser_impl.dart:6717:15)
#8      Parser.parseStatementX (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4452:14)
#9      Parser.parseStatement (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4431:20)
#10     Parser.parseIfStatement (package:_fe_analyzer_shared/src/parser/parser_impl.dart:6415:13)
#11     Parser.parseStatementX (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4468:14)
#12     Parser.parseStatement (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4431:20)
#13     Parser.parseFunctionBody (package:_fe_analyzer_shared/src/parser/parser_impl.dart:4336:15)
#14     DietListener.buildFunctionBody (package:front_end/src/fasta/source/diet_listener.dart:992:14)
#15     DietListener.endTopLevelMethod (package:front_end/src/fasta/source/diet_listener.dart:360:5)
#16     Parser.parseTopLevelMethod (package:_fe_analyzer_shared/src/parser/parser_impl.dart:2865:14)
#17     Parser.parseTopLevelMemberImpl (package:_fe_analyzer_shared/src/parser/parser_impl.dart:2641:14)
#18     Parser.parseTopLevelDeclarationImpl (package:_fe_analyzer_shared/src/parser/parser_impl.dart:502:14)
#19     Parser.parseUnit (package:_fe_analyzer_shared/src/parser/parser_impl.dart:357:15)
#20     SourceLoader.buildBody (package:front_end/src/fasta/source/source_loader.dart:461:14)
<asynchronous suspension>
#21     Loader.buildBodies (package:front_end/src/fasta/loader.dart:242:9)
<asynchronous suspension>
#22     KernelTarget.buildComponent.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:398:7)
<asynchronous suspension>
#23     withCrashReporting (package:front_end/src/fasta/crash.dart:122:12)
<asynchronous suspension>
#24     generateKernelInternal.<anonymous closure> (package:front_end/src/kernel_generator_impl.dart:164:19)
<asynchronous suspension>
#25     withCrashReporting (package:front_end/src/fasta/crash.dart:122:12)
<asynchronous suspension>
#26     compile.<anonymous closure> (package:front_end/src/api_unstable/dart2js.dart:205:37)
<asynchronous suspension>
#27     compile (package:front_end/src/api_unstable/dart2js.dart:203:35)
<asynchronous suspension>
#28     KernelLoaderTask.load.<anonymous closure> (package:compiler/src/kernel/loader.dart:182:21)
<asynchronous suspension>
#29     Compiler.runInternal (package:compiler/src/compiler.dart:279:29)
<asynchronous suspension>


#0      DietListener.buildFunctionBody (package:front_end/src/fasta/source/diet_listener.dart:1000:7)
#1      DietListener.endTopLevelMethod (package:front_end/src/fasta/source/diet_listener.dart:360:5)
#2      Parser.parseTopLevelMethod (package:_fe_analyzer_shared/src/parser/parser_impl.dart:2865:14)
#3      Parser.parseTopLevelMemberImpl (package:_fe_analyzer_shared/src/parser/parser_impl.dart:2641:14)
#4      Parser.parseTopLevelDeclarationImpl (package:_fe_analyzer_shared/src/parser/parser_impl.dart:502:14)
#5      Parser.parseUnit (package:_fe_analyzer_shared/src/parser/parser_impl.dart:357:15)
#6      SourceLoader.buildBody (package:front_end/src/fasta/source/source_loader.dart:461:14)
<asynchronous suspension>
#7      Loader.buildBodies (package:front_end/src/fasta/loader.dart:242:9)
<asynchronous suspension>
#8      KernelTarget.buildComponent.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:398:7)
<asynchronous suspension>
#9      withCrashReporting (package:front_end/src/fasta/crash.dart:122:12)
<asynchronous suspension>
#10     generateKernelInternal.<anonymous closure> (package:front_end/src/kernel_generator_impl.dart:164:19)
<asynchronous suspension>
#11     withCrashReporting (package:front_end/src/fasta/crash.dart:122:12)
<asynchronous suspension>
#12     compile.<anonymous closure> (package:front_end/src/api_unstable/dart2js.dart:205:37)
<asynchronous suspension>
#13     compile (package:front_end/src/api_unstable/dart2js.dart:203:35)
<asynchronous suspension>
#14     KernelLoaderTask.load.<anonymous closure> (package:compiler/src/kernel/loader.dart:182:21)
<asynchronous suspension>
#15     Compiler.runInternal (package:compiler/src/compiler.dart:279:29)
<asynchronous suspension>
```

**Comment by lrhn** [comment]
The code should have used `||` instead of `or`, so that's why the compiler got off tracks there, but the compiler definitely shouldn't *crash* because of a syntax error.

**Comment by johnniwinther** [comment]
cc @jensjoha

**Comment by jensjoha** [comment]
Thank you for taking the time to report this!
This was fixed in 16a66c89025ad330844a07baf76553c45cfbf7b8.
